pkg/sql/catalog/lease/count.go

```diff
@@ -15,41 +15,124 @@ import (
+	txn := db.KV().NewTxn(ctx, "count-lease")
+	executor := db.Executor()
+	err := txn.SetFixedTimestamp(ctx, at)
+	if err != nil {
+		return 0, err
+	}
+	prober := regionliveness.NewLivenessProber(db, cachedDatabaseRegions, settings)
+	regionMap, err := prober.QueryLiveness(ctx, txn)
+	if err != nil {
+		return 0, err
+	}
 	var whereClauses []string
 	for _, t := range versions {
+		versionClause := ""
+		if !forAnyVersion {
+			versionClause = fmt.Sprintf("AND version = %d", t.Version)
+		}
 		whereClauses = append(whereClauses,
-			fmt.Sprintf(`("descID" = %d AND version = %d AND expiration > $1)`,
-				t.ID, t.Version),
+			fmt.Sprintf(`("descID" = %d %s AND expiration > $1)`,
+				t.ID, versionClause),
 		)
 	}
 
-	stmt := fmt.Sprintf(`SELECT count(1) FROM system.public.lease AS OF SYSTEM TIME '%s' WHERE `,
-		at.AsOfSystemTime()) +
-		strings.Join(whereClauses, " OR ")
+	// Counts leases in non multi-region environments.
+	countLeases := func() error {
+		stmt := fmt.Sprintf(`SELECT count(1) FROM system.public.lease AS OF SYSTEM TIME '%s' WHERE `,
+			at.AsOfSystemTime()) +
+			strings.Join(whereClauses, " OR ")
+		var values tree.Datums
+		values, err = executor.QueryRowEx(
+			ctx, "count-leases", txn, /* txn */
+			sessiondata.RootUserSessionDataOverride,
+			stmt, at.GoTime(),
+		)
+		if values == nil {
+			return errors.New("failed to count leases")
+		}
+		count = int(tree.MustBeDInt(values[0]))
+		return nil
+	}
 
-	values, err := executor.QueryRowEx(
-		ctx, "count-leases", nil, /* txn */
-		sessiondata.RootUserSessionDataOverride,
-		stmt, at.GoTime(),
-	)
-	if err != nil {
-		return 0, err
+	// Counts leases by region in MR environments.
+	countLeasesByRegion := func() error {
+		stmt := fmt.Sprintf(`SELECT count(1) FROM system.public.lease AS OF SYSTEM TIME '%s' WHERE `,
+			at.AsOfSystemTime()) +
+			`crdb_region=$2::system.crdb_internal_region AND (` +
+			strings.Join(whereClauses, " OR ") +
+			")"
+		return regionMap.ForEach(func(region string) error {
+			var values tree.Datums
+			queryRegionRows := func(countCtx context.Context) error {
+				var err error
+				values, err = executor.QueryRowEx(
+					countCtx, "count-leases", txn, /* txn */
+					sessiondata.RootUserSessionDataOverride,
+					stmt, at.GoTime(), region,
+				)
+				return err
+			}
+			if hasTimeout, timeout := prober.GetTableTimeout(); hasTimeout {
+				err = timeutil.RunWithTimeout(ctx, "count-leases-region", timeout, queryRegionRows)
+			} else {
+				err = queryRegionRows(ctx)
+			}
+			if err != nil {
+				if regionliveness.IsQueryTimeoutErr(err) {
+					// Probe and mark the region potentially.
+					probeErr := prober.ProbeLiveness(ctx, region)
+					if probeErr != nil {
+						err = errors.WithSecondaryError(err, probeErr)
+						return err
+					}
+					return errors.Wrapf(err, "count-lease timed out reading from a region")
+				} else if regionliveness.IsMissingRegionEnumErr(err) {
+					// Skip this region because we were unable to find region in
```

Comment:
nit: I think we should log a warning if the enum is missing. This case should be pretty rare.